File: android/src/main/kotlin/com/example/audio_streamer/AudioStreamerPlugin.kt
package com.example.audio_streamer

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.media.audiofx.AcousticEchoCanceler
import androidx.annotation.NonNull
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.util.concurrent.Executors

class AudioStreamerPlugin : FlutterPlugin, EventChannel.StreamHandler, MethodCallHandler {
  
  private var methodChannel: MethodChannel? = null
  private val executor = Executors.newSingleThreadExecutor()
  private var eventSink: EventChannel.EventSink? = null
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    val messenger = binding.binaryMessenger
    val channel = EventChannel(messenger, "com.example.audio_streamer/events")
    channel.setStreamHandler(this)

    methodChannel = MethodChannel(messenger, "com.example.audio_streamer/methods")
    methodChannel?.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    executor.shutdownNow()
  }

  override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink) {
    this.eventSink = eventSink
    executor.execute {
      startRecording()
    }
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  private fun startRecording() {
    val sampleRate = 16000
    val channelConfig = AudioFormat.CHANNEL_IN_MONO
    val audioFormat = AudioFormat.ENCODING_PCM_16BIT
    val minBufSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat)
    val audioRecord = AudioRecord(MediaRecorder.AudioSource.VOICE_COMMUNICATION, sampleRate, channelConfig, audioFormat, minBufSize)
    var acousticEchoCanceler: AcousticEchoCanceler? = null
    // Initialize AcousticEchoCanceler for the AudioRecord instance
    if (AcousticEchoCanceler.isAvailable()) {
      acousticEchoCanceler = AcousticEchoCanceler.create(audioRecord.audioSessionId)
      acousticEchoCanceler.enabled = true
    }

    val audioData = ByteArray(minBufSize)
    audioRecord.startRecording()

    while (eventSink != null) {
      val readSize = audioRecord.read(audioData, 0, audioData.size)
      if (readSize > 0) {
        mainHandler.post {
          eventSink?.success(audioData)
        }
      }
    }
    // Cleanup
    acousticEchoCanceler?.release()
    acousticEchoCanceler = null
    audioRecord.stop()
    audioRecord.release()
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startRecording" -> {
                executor.execute {
                    startRecording()
                }
                result.success(null)
            }
            "stopRecording" -> {
                eventSink = null
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }
}

File: ios/Classes/AudioStreamerPlugin.swift
import Flutter

public class AudioStreamerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private let audioEngineManager = AudioEngineManager()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AudioStreamerPlugin()

        let methodChannel = FlutterMethodChannel(
            name: "com.example.audio_streamer/methods", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: "com.example.audio_streamer/events", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startRecording":
            if let args = call.arguments as? [String: Any],
               let recordingMode = args["recordingMode"] as? Int
            {
                audioEngineManager.startRecording(
                    recordingMode: RecordingMode(rawValue: recordingMode) ?? .normal
                ) { data in
                    self.eventSink?(data)
                }
            }
            result(nil)

        case "stopRecording":
            audioEngineManager.stopRecording()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}
